auto-update/Control/Debounce/Internal.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Control.Debounce.Internal (
+  DebounceSettings(..)
+  , DebounceEdge(..)
+  , mkDebounce
+  ) where
+
+import           Control.Concurrent      (forkIO)
+import           Control.Concurrent.MVar (takeMVar, tryPutMVar, tryTakeMVar, MVar)
+import           Control.Exception       (SomeException, handle, mask_)
+import           Control.Monad           (forever, void, when)
+import           Data.Maybe              (isJust)
+
+-- | Settings to control how debouncing should work.
+--
+-- This should be constructed using 'defaultDebounceSettings' and record
+-- update syntax, e.g.:
+--
+-- @
+-- let settings = 'defaultDebounceSettings' { 'debounceAction' = flushLog }
+-- @
+--
+-- @since 0.1.2
+data DebounceSettings = DebounceSettings
+    { debounceFreq   :: Int
+    -- ^ Microseconds lag required between subsequence calls to the debounced
+    -- action.
+    --
+    -- Default: 1 second (1000000)
+    --
+    -- @since 0.1.2
+    , debounceAction :: IO ()
+    -- ^ Action to be performed.
+    --
+    -- Note: all exceptions thrown by this action will be silently discarded.
+    --
+    -- Default: does nothing.
+    --
+    -- @since 0.1.2
+    , debounceEdge :: DebounceEdge
+    -- ^ Whether to perform the action on the leading edge or trailing edge of
+    -- the timeout.
+    --
+    -- Default: Leading.
+    --
+    -- @since 0.2.0
+    }
+
+data DebounceEdge =
+  Leading
+  -- ^ The action will either be performed immediately, or after the current cooldown period has expired.
+  | Trailing
+  -- ^ The action will be performed at the end of the current cooldown period.
+  | LeadingAndTrailing
+  -- ^ The action is performed on the leading edge. It's also performed on the trailing edge if
```

Comment:
While reviewing, I haven't looked at the implementation yet, and I'm left with a question. Let's say that the cooldown period is 2 seconds. We trigger an action, and 1 second later trigger it again. I can see two possible outcomes:

1. Wait for another 1 second, perform the action, and immediately enter the cooldown period again
2. Wait for another 1 second, perform the action, and then do not enter the cooldown period

The former seems more logical, but the latter seems to be implied. Either way, a bit more clarity in the docs would be helpful.

Reply:
The latter is how it works at the moment. So if another trigger happens at 3 seconds, then the action will happen immediately. 

The former is actually identical to the current `Leading` semantics, no?

Actually, I just played with Lodash in the console and found that for their `Leading` version they do *not* re-enter the cooldown. So if a new trigger happens at 3 seconds they perform the action immediately. This is making me wonder if there's even a difference between Lodash `Leading` and `LeadingAndTrailing`. I guess there's a subtle difference when it comes to which Javascript tick the action happens on, but that doesn't apply to us.

So really, the difference between `Leading` and `LeadingAndTrailing` for us is that `Leading` does re-enter the cooldown phase after a trigger during the previous cooldown, while `LeadingAndTrailing` does not. I guess this is a useful distinction to make--I'll try to clarify it in the docs.